### PR TITLE
Added CopyToAsync overload using default buffer size

### PR DESCRIPTION
**Pull request by khellang**
Part of https://github.com/dotnet/corefx/issues/23312

### CODE REVIEW

**Comment by khellang** [comment]
@dotnet-bot test Ubuntu x64 Checked Build and Test

File: src/mscorlib/src/System/IO/Stream.cs [hunk 1]
```diff
@@ -111,36 +111,7 @@ public virtual int WriteTimeout
...
-                    bufferSize = 1;
-                }
-                else
-                {
-                    long remaining = length - position;
-                    if (remaining > 0) // In the case of a positive overflow, stick to the default size
-                        bufferSize = (int)Math.Min(bufferSize, remaining);
-                }
-            }
+            var bufferSize = GetBufferSize(Length, Position, CanSeek);
```
Comment (stephentoub) [inline src/mscorlib/src/System/IO/Stream.cs]: Nit: var => int (same applies to the same line in the other methods)

Comment (stephentoub) [inline src/mscorlib/src/System/IO/Stream.cs]: I don't think this works.  In the original implementation, Length and Position are guarded behind the call to CanSeek; that's no longer the case.  And on most streams where CanSeek == false, Length and Position will throw.

Comment (khellang) [reply src/mscorlib/src/System/IO/Stream.cs]: Ah! I'll move them back...

**Comment by khellang** [comment]
The original code had seme comments (slightly different in the different branches). Should I leave them in?

**Comment by stephentoub** [comment]
> The original code had seme comments (slightly different in the different branches). Should I leave them in?

I went back and forth on that while reviewing your code.  They don't make as much sense when separated out into the separate method.  I'm fine with leaving them in (tweaked), I'm fine with taking them out.

**Comment by khellang** [comment]
> I went back and forth on that while reviewing your code. They don't make as much sense when separated out into the separate method. I'm fine with leaving them in (tweaked), I'm fine with taking them out.

That was my exact thinking, but wasn't sure what the tweaked version should be. If anyone has a good suggestion, I can add them back 😝

**Comment by khellang** [comment]
@dotnet-bot test Ubuntu x64 Checked Build and Test

**Comment by khellang** [comment]
@dotnet-bot test Ubuntu x64 Checked Build and Test

**Comment by stephentoub** [comment]
> If anyone has a good suggestion, I can add them back

You could say something like:
"There are no bytes left in the stream.  However, we need to ensure that CopyTo{Async} overrides are invoked so that it can provide its own validation, so we use the smallest legal buffer size."

**Comment by khellang** [comment]
I added a comment and renamed `GetBufferSize` to `GetCopyBufferSize` 😄

**Comment by stephentoub** [comment]
Before this can be exposed from corefx, it'll need to be added to corert as well.  Is that something you could do as well?

**Comment by khellang** [comment]
> Before this can be exposed from corefx, it'll need to be added to corert as well. Is that something you could do as well?

Not something I've done before, but neither is contributing to CoreCLR :wink: I'll see what I can do 😄

**Comment by khellang** [comment]
Submitted https://github.com/dotnet/corert/pull/4634 😄

**Pull request merged**